00 - Contribute to the Obsidian Hub/03 Contributor Notes/03.03 Scripts and Automation/Testing Python Code with Approval Tests.md

```diff
@@ -0,0 +1,88 @@
+# About the Python tests in [.github/scripts](https://github.com/obsidian-community/obsidian-hub/tree/main/.github/scripts)
+
+## TL;DR
+
+TODO Add a TL;DR as a quick-start...
+
+## Approval Tests in general
+
+The traditional approach for writing automated tests for code is to call a function that tests that the result of an operation matches the expectation.
+
+In pseudo code:
+
+```
+assert actual_value = expected_value
+```
+
+Approval Tests instead allow us to verify a chunk of output (such as a string or a file) in one operation as opposed to writing test assertions for each element.
+
+Complex results can be saved in a file, rather than having to do something like:
+
+- write a large number of separate asserts, for each part of the output
+- paste a large string of "expected" value in the test source code.
+
+There are Approval Tests library implementations in many different languages: [approvaltests.com](https://approvaltests.com).
+
+## Python Approval Tests
+
+[ApprovalTests.Python](https://github.com/approvals/ApprovalTests.Python)
+
+## How Approval Tests work
+
+### Approval Tests Process
+
+The simplest use of Approval Tests is to call a `verify()` method, that:
+
+1. writes out the text or object it is given to some output file (called "received")
+2. then compares the file content with previously-accepted output (called "approved").
+3. if these are equivalent, then the test passes
+4. if the received differs from the approved, or if the approved does not exist yet,
+   then the framework searches for a diff-tool on your machine,
+   and uses it to show you the differences - with:
+     - received on the left
+     - approved on the right
+
+## Dealing with test failures
+
+The diff tool allows you to easily inspect the changes between received and approved.
+
+You can then either:
+
+1. Fix the code, if the differences show an unwanted change
+2. Use the diff tool to copy the changes over from received to approved, if the changes are intentional.
+
+### Scenario: First run of a new test
+
+### Scenario: Test passes
+
+No diff tool is launched, as no action is required.
+
+### Scenario: Reviewing the output to spot the changes
+
+TODO Add screenshot and description
+
+### Scenario: Unwanted/unexpected difference
+
+TODO Add screenshot and description
+
+### Scenario: You changed the code and want to update the approved files
+
+TODO Add screenshot and description
+
+### Diff Tools (a.k.a. "Reporters")
+
+**Note**: to be used effectively, Approval Tests require you to have a diff tool installed.
```

Comment:
For Vim users, they could use vimdiff https://vi.stackexchange.com/a/626

Reply:
It's not in the list of diff tools currently supported, and right now I don't think it's trivial to use custom reporters (or rather, I haven't been able to figure out how to do so) so I think adding this might cause confusion.

Reply:
I see, I should have followed the link below. :sweat_smile: